### PR TITLE
Integration tests

### PR DESCRIPTION
**Pull request by martincostello**
Apologies in advance for the size of the diff.

This PR does _quite a lot_ of refactoring to support running the integration tests in CI using [goaws](https://github.com/p4tin/goaws).

It also _in theory_ supports running them against a real AWS account, but I haven't tested that yet.

A lot of the refactoring relates to how the region, logger factory etc. usage was spread all over the tests, making it hard to centralise into a single place to reconfigure. I've shifted a lot of things around to try and make it more-or-less changeable in a single place. This would allow, for example, a developer in Australia to run the integration tests against a real AWS account with `ap-southeast-2` as the primary region instead of `eu-west-1`.

I've also, where possible, hooked the logger factory into xunit's `ITestOutputHelper` so the logs get emitted into the test results, so you can read them if the tests fail. At a later point this could be extended into the uni tests as well.

There were also issues with tests hanging (see #380), though I've worked around these issues with usage of `ConfigureAwait(false)` in the paths for constructors that do async things.

Further, I've simplified the test project dependencies to leverage transient dependencies and reduce duplication, as well as target `netcoreapp2.1`.

Relates to #159.

### CODE REVIEW

**Comment by martincostello** [comment]
Looks like Linux containers aren't supported in the free version of AppVeyor 😞

File: README.md [hunk 1]
```diff
@@ -13,9 +13,9 @@ Before you can start publishing or consuming messages, you want to configure the
 			    new DefaultAwsClientFactory(new BasicAWSCredentials("accessKey", "secretKey"))
 ````
 
-You will also need to create a `ILoggerFactory` ([see here](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/logging)), if you do not want logging then you can always create an empty logger factory like so:
+You will also need a `ILoggerFactory` ([see here](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/logging)); if you do not want logging then you can use an empty logger factory like so:
 ````c#
-        var loggerFactory = new LoggerFactory();
+    ILoggerFactory loggerFactory = NullLoggerFactory.Instance;
```
Comment (slang25) [inline README.md]: TIL 👍

**Comment by slang25** [comment]
Yeah, the deadlocks before were due to the blocking. It was due to the xunit enforcing a test concurrency limit, this introduced a SynchronizationContext that would lock up, often the tests would pass in isolation, but a full run would almost never work.

The work I did in JustBehave was primarily a pre-req for sorting out these tests, that said, I would like to see it gone from JustSaying completely.

On goaws vs localstack. I got about 80% of tests running with localstack, cause it also emulates iam etc... That said, goaws looks much nicer, and I'd be far more likely to PR to that than localstack.

This PR rocks, let me know if I can help.

**Comment by slang25** [comment]
Shame about AppVeyor and linux images on docker ([related support question](https://help.appveyor.com/discussions/questions/8194-linux-docker-containers)), can we do it with AppVeyor linux build agent images or Travis?

**Comment by martincostello** [comment]
My initial thinking yesterday was maybe we could run goaws in Travis (it's working now 🎉) and once I've done some local testing with it, we could setup AppVeyor to do the "real" integration tests as a "workaround".

Then we've got PRs testing both with the local emulator on Linux and against real SNS and SQS on Windows?

It's not ideal that we've got different coverage on different OSs, but it would spread things around, at least for an initial merge.

My initial aim was to get this at least green and in a decent working state for a jumping off point for merging to master, then we could do any further improvements on top of that (extending coverage, maybe removing JustBehave etc.).

**Comment by martincostello** [comment]
Fun fact: dynamically changing the build Id in the build makes the pending GitHub status for AppVeyor 404:

```
Expected: https://ci.appveyor.com/project/justeattech/justsaying/build/0.0.386-utfjxwpu
Actual: https://ci.appveyor.com/project/justeattech/justsaying/build/6.0.0-beta2-build386-d5lx2cw2
```

**Comment by martincostello** [comment]
I've got far enough along in this PR I think that there's enough value there to merge it now even though the overall task in #159 isn't yet complete.

Merging this adds integration tests with goaws for Travis CI, plus locally via opt-in in `Build.ps1` if you have Docker installed for Linux containers.

In a follow-up PR I'll do work to ensure we also support running the same test suite against AWS itself if the appropriate environment variables for keys etc. are configured.

**Comment by martincostello** [comment]
This also gets enough of the refactoring into master to allow others to work on #380.

**Comment by slang25** [comment]
This is amazing! 🎉 I bought Paul a coffee to say thanks for goaws

**Pull request merged**